### ISSUE TITLE
Null-aware return doesn't promote variable

### ISSUE THREAD

**Issue by feinstein** [dart-lang/language#4156]
Consider this code:

```dart
final lastTransaction = transactions.lastOrNull;
if (lastTransaction?.expirationDate == null) {
  return false;
}

final expirationDate = DateTime.tryParse(lastTransaction!.expirationDate!);
```

If I remove the `!`, I get an analyser error saying I can't use `lastTransaction` because it can be `null`. I believe the analyser should be able to conclude that `lastTransaction` can't be `null`, because it's a final local variable, and the `?.` in the `if` check would ensure there's no valid code path for it to still be `null`.

**Comment by lrhn** [comment]
Promotion does not depend on a getter coming from a final variable.

Any non-constant getter is assumed to be able to give a different value the next time it's read, so it cannot be promoted.
(Only exceptions are final _library local_ variables that are not overridden with a getter inside the same library. The compiler trusts that there are no overrides elsewhere, and that the author and the user agrees that this is a final variable.)

If the compiler allowed promotion of final variables from other libraries, it would make it a breaking change to change a final variable to a getter, even if it doesn't change the value. Allowing that changed is one of the points of having getters.

**Comment by mateusfccp** [comment]
Similar to #1224.

**Comment by feinstein** [comment]
> In this case x?.y != null can promote x to non- null, but x?.y == null does not imply that x is null.

@lrhn I agree that `x?.y == null` doesn't imply that x is null, but in this case the compiler could be able to have some more sofisticated logic to promote it, as if the if-check is not null, then x could not be null.

Do you want me to leave this as a request or close it?

@mateusfccp I made it work, thanks, my goal with this snippet was more to showcase my point.
